bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1040,18 +1045,83 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
     }
 
     private static boolean checkIsTupleType(BType sourceType, BTupleType targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.TUPLE_TAG) {
+
+        if (sourceType.getTag() == TypeTags.UNION_TAG) {
+            for (BType memberType : ((BUnionType) sourceType).getMemberTypes()) {
+                if (!checkIsTupleType(memberType, targetType, unresolvedTypes)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+            return false;
+        }
+
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        BType targetRestType = targetType.getRestType();
+
+        BTupleType sourceTupleType;
+        if (sourceType.getTag() == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case UNSEALED:
+                    if (targetRestType == null) {
+                        return false;
+                    } else {
+                        return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                case CLOSED_SEALED:
+                    if (sourceArrayType.getSize() >= targetTypes.size()) {
+                        if (targetTypes.isEmpty()) {
+                            if (targetRestType != null) {
+                                return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                            }
+                            return true;
+                        }
+
+                        boolean isSameType = true;
+
+                        for (BType targetElementType : targetTypes) {
+                            if (!(checkIsType(sourceElementType, targetElementType, unresolvedTypes))) {
+                                isSameType = false;
+                            }
+                        }
+                        if (sourceArrayType.getSize() > targetTypes.size()) {
```

Comment:
We can reduce the indentation a bit by doing
```suggestion
                        if (sourceArrayType.getSize() == targetTypes.size()) {
                            return isSameType;
                        }
                        if (targetRestType != null) {
                        ...
```